### PR TITLE
Proposal: ICRC-2 Standard Specification (Approve / Transfer From)

### PR DESCRIPTION
**Pull request by ozwaldorf**
The proposal for the `ICRC-2` Approve/Transfer From extension to the base `ICRC-1` standard!

https://github.com/Psychedelic/ICRC-1/tree/main/extensions/ICRC-2

### CODE REVIEW

**Comment by ozwaldorf** [comment]
I created a folder named `extensions` as discussed in the WG for the initial layout, but I think reorganizing the repository further would be beneficial. Now that we have extensions, I think a more flat layout would be better, with each ICRC-X having its dedicated folder, with the spec, did, reference implementations, etc.

```
ICRC/
  - Standards/
    - ICRC-1/
      - ref/
      - spec.did
      - SPEC.md
    - ICRC-2/
  - README.md
  - LICENSE
  - etc...
```

**Comment by ozwaldorf** [comment]
I created a folder named `extensions` as discussed in the WG for the initial layout, but I think reorganizing the repository further would be beneficial. Now that we have extensions, I think a more flat layout would be better, with each ICRC-X having its dedicated folder, with the spec, did, reference implementations, etc.

```
ICRC/
  - Standards/
    - ICRC-1/
      - ref/
      - interface.did
      - README.md
    - ICRC-2/
      - ref
      - interface.did
      - README.md
  - README.md
  - LICENSE
  - etc...
```

**Comment by roman-kashitsyn** [comment]
I like the idea of having the `standards` directory!

> with each ICRC-X having its dedicated folder, with the spec, did, reference implementations, etc.

I would keep one reference implementation that implements all the extensions; that will be a great test that these implementations play well together. The test suite can also be monolithic and inspect the `supported_standard` to skip irrelevant test cases.

**Comment by roman-kashitsyn** [comment]
Thanks a lot for your proposal, @ozwaldorf!
I proposed a few refinements and improvements in https://github.com/Psychedelic/ICRC-1/pull/1; I'd appreciate it if you could take a look.

File: extensions/ICRC-2/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    fee : opt nat;
+    memo : opt blob;
+    created_at_time : opt nat64;
+};
+
+type TransferFromArgs = record {
+    from : Account;
+    to : Account;
+    amount : nat;
+    fee : opt nat;
```
[the remark this answers is not in the document]
Comment (ozwaldorf) [reply extensions/ICRC-2/README.md:112]: Agreed, I think the approver's balance should be the source, as the caller isn't guaranteed to have funds

File: extensions/ICRC-2/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+# `ICRC-2`: Approve and Transfer From
+
+## Abstract
+
+`ICRC-2` is an extension to the base `ICRC-1` standard. It adds the ability to approve a third party, and then that third party can transfer tokens on your behalf.
+
+The approve and transfer-from flow is a 2-step process.
+1. Signer A approves X tokens for signer b
+2. Signer B can transfer on behalf of signer A, up to the allowance X, or Signer B’s total balance
```
[the remark this answers is not in the document]
Comment (ozwaldorf) [reply extensions/ICRC-2/README.md]: Good catch! Of course it's fixed with the PR, but I had meant signer A :facepalm:

**Comment by ozwaldorf** [comment]
Merged in your changes, everything looked great

One uncertainty I have is how memo and created_at_time would be implemented. Would utilizing unique memos result in separate allowances in the ledger state? The transfer memo works good, as the memo ends up in the tx and doesn't have bearing on the canister state, but this would, right?

**Comment by ozwaldorf** [comment]
> I like the idea of having the `standards` directory!
> 
> > with each ICRC-X having its dedicated folder, with the spec, did, reference implementations, etc.
> 
> I would keep one reference implementation that implements all the extensions; that will be a great test that these implementations play well together. The test suite can also be monolithic and inspect the `supported_standard` to skip irrelevant test cases.

Should this PR provide said changes? Or should it be done separately after

**Comment by roman-kashitsyn** [comment]
> The transfer memo works good, as the memo ends up in the tx and doesn't have bearing on the canister state, but this would, right?

I think we need a new TX type for Approve transactions. We can use memo and created_at_time to deduplicate requests, put them into a TX metadata, and ignore them in the approval map, just as we ignore memo in the balance map.

> Would utilizing unique memos result in separate allowances in the ledger state?

Memos have two purposes:
* Deduplication in case of retries in case of a networking failure.
* As a correlation identifier if you want to link the TX to an Invoice ID (requires block access).

> Should this PR provide said changes? Or should it be done separately after

We can do it in a separate PR, too much work to do everything in one go.
It might also be helpful to figure out the details with the WG before rushing into an implementation.

**Comment by roman-kashitsyn** [comment]
My main concern about the original approve/transferFrom flow is that it does not play well with asynchrony:

> The flow does not support concurrency: if I execute two payments to the same shop asynchronously, only one of the payments will likely succeed (the exact outcome depends on the message scheduling order).
-- https://mmapped.blog/posts/10-payment-flows.html#approve-transfer-from

This paper https://swcregistry.io/docs/SWC-114 describes a related problem.

One remediation the paper suggests is implementing a CAS scheme, where each approval requires the previously approved value to succeed.

Another direction we could explore is making approvals additive (proposed here https://github.com/dfinity/ICRC-1/issues/22). 
E.g., when Alice approves five tokens to Bob and then approves ten tokens to Bob, the total approval is 15.
To remove the approval, Alice approves –15 tokens (we'll have to make the approval signed).

**Pull request merged**